templates/al2023/runtime/gpu/nvidia-kmod-load.sh

```diff
@@ -0,0 +1,75 @@
+#!/usr/bin/env bash
+
+set -o errexit
+set -o pipefail
+set -o nounset
+
+if ! gpu-ami-util has-nvidia-devices; then
+  echo >&2 "no NVIDIA devices are present, not loading kernal module!"
+  return 0
+fi
+
+NVIDIA_VENDOR_ID="10de"
+PCI_CLASS_CODES=(
+  "0300" # VGA controller; instance types like g3, g4
+  "0302" # 3D controller; instance types like p4, p5
+)
+
+INSTANCE_TYPE=$(imds /latest/meta-data/instance-type)
+
+# return the path of the file containing devices supported by the nvidia-open kmod
+# fail if the expected file doesn't exist
+function nvidia-open-supported-devices-file() {
+  local KMOD_MAJOR_VERSION=$(rpmquery kmod-nvidia-latest-dkms --queryformat '%{VERSION}' | cut -d. -f1)
+  local SUPPORTED_DEVICE_FILE="/etc/eks/nvidia-open-supported-devices-${KMOD_MAJOR_VERSION}.txt"
+  if ! test -f "${SUPPORTED_DEVICE_FILE}"; then
+    echo >&2 "Supported device file not found for ${KMOD_MAJOR_VERSION}: ${SUPPORTED_DEVICE_FILE}"
+    exit 1
+  fi
+  echo "${SUPPORTED_DEVICE_FILE}"
+}
+
+# determine if the attached nvidia devices are supported by the open-source kernel module
+function devices-support-open() {
+  local SUPPORTED_DEVICE_FILE=$(nvidia-open-supported-devices-file)
+  for PCI_CLASS_CODE in "${PCI_CLASS_CODES[@]}"; do
+    for NVIDIA_DEVICE_ID in $(lspci -n -mm -d "${NVIDIA_VENDOR_ID}::${PCI_CLASS_CODE}" | awk '{print $4}' | tr -d '"' | tr '[:lower:]' '[:upper:]'); do
+      if ! grep "^0x${NVIDIA_DEVICE_ID}\s" "${SUPPORTED_DEVICE_FILE}"; then
+        return 1
+      fi
+    done
+  done
+  return 0
+}
+
+# load the nvidia kernel module appropriate for the attached devices
+MODULE_NAME=""
+if devices-support-open; then
```

Comment:
since the older cards that require the proprietary driver are a known quantity, and new cards that will be released in the future are unknown to this script (and therefore the PCI class codes are unknown here) - should we flip this logic to default to using the open GPU kernel module unless we know the proprietary driver is absolutely required?

Reply:
Let's check with @Issacwww about that. I'm not the original writer of this.

Reply:
I think this is a fair call out, I will follow up with next PR as current one is big enough :)

Reply:
Sounds good, I'm also in favor of changing this in different PR.

Reply:
> new cards that will be released in the future are unknown to this script (and therefore the PCI class codes are unknown here)

The PCI class codes don't change per device, these have been the same for a few decades now. :)

> should we flip this logic to default to using the open GPU kernel module unless we know the proprietary driver is absolutely required

I didn't go this route because it's way messier. Compiling an exhaustive list of every PCI device ID used by those EC2 instance types is quite a chore. With the current approach, we generate the supported device file from the driver archive and do a simple lookup.

Long term, yes we want to default to the open source kmod. NVIDIA is going to do that in the 560 driver branch, and we can revisit/refactor accordingly then, IMO.